crates/pico/src/memo.rs

```diff
@@ -0,0 +1,180 @@
+            }
+            time_calculated
+        } else {
+            db.storage()
+                .nodes()
+                .get(&node_id)
+                .expect("node should exist. This is indicative of a bug in Isograph.")
+                .time_calculated
+        }
+    } else {
+        let (value, dependencies, time_calculated) =
+            call_inner_fn_and_collect_dependencies(db, param_id, inner_fn);
+        let current_epoch = db.current_epoch();
+        db.storage_mut().nodes().insert(
+            node_id,
+            DerivedNode {
+                time_verified: current_epoch,
+                time_calculated,
+                dependencies,
+                inner_fn,
+            },
+        );
+        db.storage_mut().values().insert(node_id, value);
+        time_calculated
+    };
+    register_dependency(db, NodeKind::Derived(node_id), time_calculated);
+    node_id
+}
+
+fn any_dependency_changed<Db: Database>(db: &mut Db, node_id: &NodeId) -> bool {
+    let current_epoch = db.current_epoch();
+    let dependencies = db
+        .storage()
+        .nodes()
+        .get(node_id)
+        .expect("node should exist. This is indicative of a bug in Isograph.")
+        .dependencies
+        .iter()
+        .filter_map(|dep| {
+            if dep.time_verified_or_calculated != current_epoch {
+                Some((dep.node_to, dep.time_verified_or_calculated))
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    dependencies
+        .into_iter()
+        .any(|(dep_node, time_verified)| match dep_node {
+            NodeKind::Source(key) => source_node_changed(db, &key, time_verified),
+            NodeKind::Derived(dep_node_id) => derived_node_changed(db, &dep_node_id),
+        })
+}
+
+fn source_node_changed<Db: Database>(db: &Db, key: &SourceKey, time_verified: Epoch) -> bool {
+    match db.storage().sources().get(key) {
+        Some(source) => source.time_calculated > time_verified,
+        None => true,
+    }
+}
+
+fn derived_node_changed<Db: Database>(db: &mut Db, node_id: &NodeId) -> bool {
+    if !db.storage().params().contains_key(&node_id.param_id) {
+        return true;
+    }
+    let mut has_changed = false;
+    let inner_fn = db
+        .storage()
+        .nodes()
+        .get(node_id)
+        .expect("node should exist. This is indicative of a bug in Isograph.")
+        .inner_fn;
+
+    let (new_value, dependencies, time_calculated) =
+        call_inner_fn_and_collect_dependencies(db, node_id.param_id, inner_fn);
+    if let Some(value) = db.storage_mut().values().get_mut(node_id) {
+        if *value != new_value {
+            *value = new_value;
+            has_changed = true;
+        }
+    } else {
+        db.storage_mut().values().insert(*node_id, new_value);
+        has_changed = true;
+    }
+    let current_epoch = db.current_epoch();
+    if let Some(node) = db.storage_mut().nodes().get_mut(node_id) {
```

Comment:
why do we call `get_mut` twice here? Can we return a mutable reference from the above if block, or move the `node.dependencies = ...` into both branches of the if block?